File: lecture_011(RecursionBacktracking)/l_001_floodfill.java
import java.util.Scanner;
public class l_001_floodfill {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        int n = scn.nextInt();
        int m = scn.nextInt();
        int[][] arr = new int[n][m];
        for(int i=0; i<arr.length; i++) {
            for(int j=0; j<arr[0].length; j++) {
                arr[i][j] = scn.nextInt();
            }
        }
        boolean[][] visited = new boolean[n][m];
        floodFill(arr,0,0,"",visited);
    }
    public static void floodFill(int[][] maze, int row, int col, String path, boolean[][] visited) {
        if(row<0 || col<0 || row==maze.length || col==maze[0].length || 
        maze[row][col]==1 || visited[row][col]==true) {
            return;
        }
        if(row==maze.length-1 && col==maze[0].length-1) {
            System.out.println(path);
            return;
        }
        visited[row][col] = true;
        floodFill(maze,row-1,col,path+"t",visited);     //top
        floodFill(maze,row,col-1,path+"l",visited);     //left
        floodFill(maze,row+1,col,path+"d",visited);     //down
        floodFill(maze,row,col+1,path+"r",visited);     //right
        visited[row][col] = false;
    }
}

File: lecture_011(RecursionBacktracking)/l_001_floodfill.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

void floodfill(vector<vector<int>> &maze, int row, int col, string path, vector<vector<bool>> &visited) {
    if(row<0 || col<0 || row==maze.size() || col==maze[0].size() || 
    maze[row][col]==1 || visited[row][col]==true) {
        return;
    }
    if(row==maze.size()-1 && col==maze[0].size()-1) {
        cout<<path<<endl;
        return;
    }
    visited[row][col] = true;
    floodfill(maze,row-1,col,path+"t",visited);     //top
    floodfill(maze,row,col-1,path+"l",visited);     //left
    floodfill(maze,row+1,col,path+"d",visited);     //down
    floodfill(maze,row,col+1,path+"r",visited);     //right
    visited[row][col] = false;
}
int main() {
    int n,m;
    cin>>n,m;
    vector<vector<int>>arr(n,vector<int>(m,0));
    for(int i=0; i<arr.size(); i++) {
        for(int j=0; j<arr[0].size(); j++) {
            cin>>arr[i][j];
        }
    }
    vector<vector<bool>>visited(n,vector<bool>(m,0));
    floodfill(arr,0,0,"",visited);

    return 0;
}